### ISSUE TITLE
runscript is being wiped out during build process

### ISSUE THREAD

**Issue by cmpetty** [apptainer/apptainer#2561]
### Version of Apptainer

```
apptainer version 1.3.4-1.el9
```

### Expected behavior

Expect to see a valid build / executable sif image

### Actual behavior

During the build process, the runscript is being wiped out 

```
(base) [cmp12@edison singularity]$ singularity run fmriprep-v24.1.0.sif
INFO:    Environment variable SINGULARITY_DISABLE_CACHE is set, but APPTAINER_DISABLE_CACHE is preferred
There is no runscript defined for this container
```

Tree of the .singularity.d directory within the sandbox
```
(base) [cmp12@edison singularity]$ tree fmriprep_v24.1.0/.singularity.d/
fmriprep_v24.1.0/.singularity.d/
├── actions
│   ├── exec
│   ├── run
│   ├── shell
│   ├── start
│   └── test
├── env
│   ├── 01-base.sh
│   ├── 10-docker2singularity.sh
│   ├── 90-environment.sh
│   ├── 94-appsbase.sh
│   ├── 95-apps.sh
│   ├── 99-base.sh
│   └── 99-runtimevars.sh
├── labels.json
├── libs
├── runscript
├── Singularity
└── startscript
```

Contents of the runscript are definitely not empty in the sandbox, which executes correctly as a sandbox.  I've abbreviated the full runscript for brevity, but it is present in the sandbox which functions normally:

```
#!/bin/bash
OCI_ENTRYPOINT='"/opt/conda/envs/fmriprep/bin/fmriprep"'
OCI_CMD=''

# When SINGULARITY_NO_EVAL set, use OCI compatible behavior that does
# not evaluate resolved CMD / ENTRYPOINT / ARGS through the shell, and
# does not modify expected quoting behavior of args.
if [ -n "$SINGULARITY_NO_EVAL" ]; then
    # ENTRYPOINT only - run entrypoint plus args
    if [ -z "$OCI_CMD" ] && [ -n "$OCI_ENTRYPOINT" ]; then
        set -- '/opt/conda/envs/fmriprep/bin/fmriprep' "$@"

        exec "$@"
    fi

    # CMD only - run CMD or override with args
    if [ -n "$OCI_CMD" ] && [ -z "$OCI_ENTRYPOINT" ]; then
        exec "$@"
    fi
```

### Steps to reproduce this behavior

How can others reproduce this issue/problem?

```
#set some envs
export APPTAINER_DISABLE_CACHE=true
export APPTAINER_PULLFOLDER=`pwd`
export APPTAINER_CACHEDIR=`pwd`
export TMPDIR=`pwd`
```

Pull the image from docker as a sandbox:
```
singularity --verbose build --sandbox fmriprep_v24.1.0/ docker://nipreps/fmriprep:latest
```

After adding a single file, not related to apptainer, rebuild it.

```
singularity --verbose build fmriprep-v24.1.0.sif fmriprep_v24.1.0/
```

The runscript in the build sif now contains:
```
(base) [cmp12@edison singularity]$ singularity shell fmriprep-v24.1.0.sif
INFO:    Environment variable SINGULARITY_DISABLE_CACHE is set, but APPTAINER_DISABLE_CACHE is preferred
Apptainer> cat /.singularity.d/runscript
#!/bin/sh
# Copyright (c) Contributors to the Apptainer project, established as
#   Apptainer a Series of LF Projects LLC.
#   For website terms of use, trademark policy, privacy policy and other
#   project policies see https://lfprojects.org/policies
# Copyright (c) 2018-2021, Sylabs Inc. All rights reserved.
# This software is licensed under a 3-clause BSD license. Please consult the
# LICENSE.md file distributed with the sources of this project regarding your
# rights to use or distribute this software.

echo "There is no runscript defined for this container\n";
```

### What OS/distro are you running

```Red Hat Enterprise Linux release 9.4 (Plow)```

### How did you install Apptainer

```dnf install apptainer```

If i pull the image directly from docker, without the sandbox, this is not happening, but i need to add files, which i've always done.

```singularity pull docker://nipreps/fmriprep:latest```

**Comment by DrDaveD** [comment]
Thanks for the report.  Indeed, any added files in /.singularity.d are successfully copied, but any of the standard files are overwritten.  Since it isn't copying from an oci source at that time, it puts in that default runscript.  It should probably be changed to preserve at least the runscript and the startscript if they already exist.

**Comment by ikaneshiro** [comment]
So this looks like an oversight during the rewrite of Singularity 2.x to 3.x in Go. It looks like the intended behavior of local build sources is to [preserve runscripts](https://github.com/apptainer/singularity/pull/834#issuecomment-318199520) from the base image, but we currently always truncate the file regardless if a definition file specifies one or not.

**Comment by cmpetty** [comment]
i will add that this behavior seems to be a more recent phenomenon.

The last images i built are dated 04/12/2024, those were with singularity-ce, which would've been 4.#.#.  Exactly the same procedure for building.

I switched to apptainer after the new build failed via singularity-ce with the same errors as above.  Have the same outcome with either.

**Comment by DrDaveD** [comment]
It's true.  I tried a bunch of older versions of apptainer and it was OK all the way through 1.3.3.  It was a bug introduced in 1.3.4.  None of the PR titles sound like they'd be relevant, so we need to dig into each one.

If singularity-ce has recently introduced the same problem, perhaps it was caused by something we imported from there or they imported from us.  I suggest reporting the problem there too.

**Comment by DrDaveD** [comment]
I suspect  #2397.  It changed the local image packer which is the affected code.  The same PR was also submitted to singularity-ce.

**Comment by DrDaveD** [comment]
Yes, because it inserted a late call to [makeBaseEnv()](https://github.com/apptainer/apptainer/blame/release-1.3/internal/pkg/build/sources/conveyorPacker_local.go#L127).  

@Flamefire please take a look.

**Comment by Flamefire** [comment]
Indeed, the issue is https://github.com/apptainer/apptainer/blob/259c4621881193f3591b7e3dfd278d744b472be1/internal/pkg/build/sources/base_environment.go#L159-L161

That doesn't make sense to me: Previously there wasn't supposed to be any "base environment", i.e. files/folders. So truncating the file isn't useful. After the change files might exist and truncating them leads to this issue.

I would suggest to just fix the permissions if the file exists and don't overwrite it. But I'm not sure if there has been a reason for the overwriting as the behavior has [always been there](https://github.com/apptainer/apptainer/commit/55e18021287a204d16b35a88dbf2033ade6c2b36#diff-9827a5f5e04eab2640d7c406266e394f721f59bb9adbbc1ab4434426457b4fb7), even before considering that the file(s) might exist (and need fixed permissions)

**Comment by DrDaveD** [comment]
It was caused by changing the order, right?  Previously `makeBaseEnv` was called before `GetLocalPacker`, so presumably that means the original files overwrote whatever `makeBaseEnv` created.  Now `makeBaseEnv` is overwriting the original files.

**Comment by DrDaveD** [comment]
Nevermind my last comment.  You were already implicitly acknowledging that but I didn't catch that at first.

**Comment by DrDaveD** [comment]
There's a fix at sylabs/singularity#3362 we could import.  @JasonYangShadow can you do it?
